### PR TITLE
Split some classes from Kokkos_ViewMapping

### PR DESCRIPTION
**Pull request by nmm0**
This pulls some classes out from Kokkos_ViewMapping in a separate header. This is required for https://github.com/kokkos/kokkos/pull/6830 (and partitions the PRs a bit).

This should also hopefully untangle the View Mapping code slightly.

### CODE REVIEW

**Review by ldh4 — APPROVED** [review]

**Comment by nmm0** [comment]
retest this please

File: core/src/impl/Kokkos_ViewDataAnalysis.hpp [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+
+#ifndef KOKKOS_IMPL_PUBLIC_INCLUDE
+static_assert(false,
+              "Including non-public Kokkos header files is not allowed.");
+#endif
+
+#ifndef KOKKOS_VIEW_DATA_ANALYSIS_HPP
+#define KOKKOS_VIEW_DATA_ANALYSIS_HPP
+
+#include <Kokkos_Core_fwd.hpp>
```
Comment (dalg24) [inline core/src/impl/Kokkos_ViewDataAnalysis.hpp:25]: Why include this header rather than just `<Kokkos_Macros.hpp>`?
What else is being used here?

**Review by dalg24 — COMMENTED** [review]
Please confirm that you didn't change anything besides the `Kokkos::Impl` nested namespace definition

**Comment by dalg24** [comment]
> Please confirm that you didn't change anything besides the `Kokkos::Impl` nested namespace definition

I diffed it and indeed no code changes besides it.
I would prefer if you changed the header include.

File: core/src/impl/Kokkos_ViewDataAnalysis.hpp (hunk 1 above)
Comment (dalg24) [reply core/src/impl/Kokkos_ViewDataAnalysis.hpp:25]:
```suggestion
#include <Kokkos_Macros.hpp>
```

**Review by masterleinad — APPROVED** [review]

**Comment by nmm0** [comment]
> > Please confirm that you didn't change anything besides the `Kokkos::Impl` nested namespace definition
> 
> I diffed it and indeed no code changes besides it. I would prefer if you changed the header include.

Yep, just moved everything no changes to the code

**Review by crtrott — APPROVED** [review]

**Pull request merged**